x/smartaccount/ante/auth.go

```diff
@@ -0,0 +1,285 @@
+package ante
+
+import (
+	"encoding/json"
+	"fmt"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/crypto/types/multisig"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/cosmos/cosmos-sdk/types/tx/signing"
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	authsigning "github.com/cosmos/cosmos-sdk/x/auth/signing"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	"github.com/terra-money/core/v2/x/smartaccount/types"
+)
+
+// SmartAccountAuthDecorator does authentication for smart accounts
+type SmartAccountAuthDecorator struct {
+	sak                       SmartAccountKeeper
+	wk                        WasmKeeper
+	ak                        authante.AccountKeeper
+	signModeHandler           authsigning.SignModeHandler
+	defaultVerifySigDecorator sdk.AnteHandler
+}
+
+func NewSmartAccountAuthDecorator(
+	sak SmartAccountKeeper,
+	wk WasmKeeper,
+	ak authante.AccountKeeper,
+	sigGasConsumer func(meter sdk.GasMeter, sig signing.SignatureV2, params authtypes.Params) error,
+	signModeHandler authsigning.SignModeHandler,
+) SmartAccountAuthDecorator {
+	if sigGasConsumer == nil {
+		sigGasConsumer = authante.DefaultSigVerificationGasConsumer
+	}
+	defaultVerifySigDecorator := sdk.ChainAnteDecorators(
+		authante.NewSetPubKeyDecorator(ak),
+		authante.NewValidateSigCountDecorator(ak),
+		authante.NewSigGasConsumeDecorator(ak, sigGasConsumer),
+		authante.NewSigVerificationDecorator(ak, signModeHandler),
+	)
+	return SmartAccountAuthDecorator{
+		sak:                       sak,
+		wk:                        wk,
+		ak:                        ak,
+		signModeHandler:           signModeHandler,
+		defaultVerifySigDecorator: defaultVerifySigDecorator,
+	}
+}
+
+// AnteHandle checks if the tx provides sufficient fee to cover the required fee from the fee market.
+func (sad SmartAccountAuthDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// check if the tx is from a smart account
+	sigTx, ok := tx.(authsigning.SigVerifiableTx)
+	if !ok {
+		return ctx, sdkerrors.ErrInvalidType.Wrap("expected SigVerifiableTx")
+	}
+
+	// Signer here is the account that the state transition is affecting
+	// e.g. Account that is transferring some Coins
+	signers := sigTx.GetSigners()
+	// Current only supports one signer (TODO review in the future)
+	if len(signers) != 1 {
+		return ctx, sdkerrors.ErrorInvalidSigner.Wrap("only one account is supported (sigTx.GetSigners()!= 1)")
+	}
+	account := signers[0].String()
+
+	setting, err := sad.sak.GetSetting(ctx, account)
```

Comment:
Should we move this check all the way to the top of the method? Since the default auth ante handler supports multiple signers (?), we don't enforce this check unless it is a smart account